### ISSUE TITLE
MAINT: Warning: Fortran 2018 deleted feature: Shared DO termination

### ISSUE THREAD

**Issue by oliver-schumann** [scipy/scipy#19557]
### Describe your issue.

While writing a cpp wrapper for scipy, the following warning was shown due to the very old Fortran coding style used in the fitpack part in interpolation.
```
interpolate/fitpack/*.f
```

Warning: Fortran 2018 deleted feature: Shared DO termination

This can easily be solved. E.g.:

```
      do 20 i=1,k1
         t(i) = xb
         t(j) = xe
         j = j-1
  20  continue
  ```
  
must be changed to 

```
      do i=1,k1
         t(i) = xb
         t(j) = xe
         j = j-1
      end do
```

### Reproducing Code Example

```python
none
```

### Error message

```shell
Warning: Warning: Fortran 2018 deleted feature: Shared DO termination
```

### SciPy/NumPy/Python version and system information

```shell
1.11.1 1.24.4 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.10.4
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.35
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-hy2eo89f/overlay/lib/python3.10/site-packages/pythran
    version: 0.13.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

**Comment by lucascolley** [comment]
This is in `interpolate/fitpack/curfit.f`.

https://github.com/scipy/scipy/blob/86ab60469451c3bfb8c3e5f150c862b3c50e4f42/scipy/interpolate/fitpack/curfit.f#L241C19-L245

**Comment by ev-br** [comment]
The language feature is ugly indeed! That said, it is unlikely to be really removed from the language / compiler support. 
So the warning is harmless is somewhat annoying. It's probably easiest to silence the warning during the build.

We can in principle review a PR which "modernizes" these do-enddo loops, but --- If you're up to looking into Fortran code carried by scipy, you're most welcome to join the effort in https://github.com/scipy/scipy/issues/18566; if you're up to looking into FITPACK specifically, there's https://github.com/scipy/scipy/issues/2579.

Both these issues track efforts where help is wanted and very much appreciated; both are likely to bring more bang for the effort.

**Comment by ivan-pi** [comment]
The code snippet you've shown, and that @lucascolley has linked, has no shared termination.

Shared termination refers to the case where nested loops end with the same continue statement:

```fortran
C shared_do.f
C
      do 10 i = 1, 3
      do 10 j = 1, 3
      write(*,*) i, j
c >>> SHARED TERMINATION <<<
   10 continue

      do 40 i = 1, 3
      do 50 j = 1, 3
      write(*,*) i, j
   50 continue
   40 continue
      end   
```

Only the first case will raise a warning:
```
$ gfortran shared_do.f 
shared_do.f:4:72:

    4 |       do 10 j = 1, 3
      |                                                                        1
Warning: Fortran 2018 deleted feature: Shared DO termination label 10 at (1)
```

So if the warning bothers you, you can fix it relatively easily by adding an explicit continue statement. The labelled do loop remains valid F2018 syntax (albeit marked as obsolescent).

**Comment by czgdp1807** [comment]
One suggestion - I can write a simple script to convert all such shared do terminations to `do-endo` combinations. Doing the same thing manually would be hard and time consuming. Like if we need to get past these warnings then I can do it.

Let me know. I will start on it.

**Comment by ev-br** [comment]
If anything, if I were doing it, I'd much rather do things manually. That said, as long as the end result is the same, whichever is easier for who is doing it.

Whether this is worth spending effort on, I don't know. Is there a Fortran compiler which has difficulties with it?
We filter these warnings anyway, they've caused no issues for many years, given how conservative Fortran compilers are, I doubt they'll graduate a warning into an error in a foreseeable future. 

To summarize, I'm +/-0: if someone's motivated enough, great, let's fix these.

**Comment by rgommers** [comment]
Ah, I misread the original issue report. We are indeed already passing `-std=legacy` for GCC, and the original bug report does use GCC. So it seems like the "While writing a cpp wrapper for scipy" part also involves using a Fortran compiler, and that was being done without `-std=legacy`.

Hence I'd vote for doing nothing here and closing this issue. This Fortran code is going to be removed at some point, so no need to spend much time on it now.
